### PR TITLE
Client optimizations

### PR DESCRIPTION
**Pull request by redknightlois**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-16213
https://issues.hibernatingrhinos.com/issue/RavenDB-16207
https://issues.hibernatingrhinos.com/issue/RavenDB-16202

### Type of change
- Optimization

### How risky is the change?
- Low 

### Backward compatibility
- Non breaking change

### CODE REVIEW

**Comment by redknightlois** [comment]
These are a few commits to improve a little bit here and there, but I wouldn't call them solved by any means. There is not much we can improve without pushing onto how things work at the client level. There is almost no low hanging fruits left without very specific high pain examples we can use to guide improvements.

File: src/Sparrow/Utils/WeakTypeCache.cs [hunk 1]
```diff
@@ -0,0 +1,102 @@
+﻿using System;
+using System.Buffers;
+using System.Threading;
+using Sparrow.Binary;
+
+namespace Sparrow.Utils
+{
+    public class WeakTypeCache<T> : IDisposable
```
Comment (arekpalinski) [inline src/Sparrow/Utils/WeakTypeCache.cs]: I don't see any usage of this class. Was it supposed to be used somewhere?

File: bench/Micro.Benchmark/Benchmarks/TypeCache.cs [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+            public bool TryGet(Type t, out T value)
+            {
+                unsafe
+                {
+                    // PERF: This only works because we assume that Type will not change the memory location over time.
+                    // If that would not be the case we would be causing lots of cache misses, which is not the intended
+                    // use of this facility. The idea is to be incredibly fast to retrieve a value if it exist in the cache
+                    // even at the expense of failing to recognize that it is there.
+                    TypedReference reference = __makeref(t);
+                    ulong extremelyUnsafeAddress = **(ulong**)&reference;
```
Comment (arekpalinski) [inline bench/Micro.Benchmark/Benchmarks/TypeCache.cs:171]: Please check the warning

File: bench/Micro.Benchmark/Benchmarks/TypeCache.cs [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+            public void Put(Type t, T value)
+            {
+                unsafe
+                {
+                    // PERF: This only works because we assume that Type will not change the memory location over time.
+                    // If that would not be the case we would be causing lots of cache misses, which is not the intended
+                    // use of this facility. The idea is to be incredibly fast to retrieve a value if it exist in the cache
+                    // even at the expense of failing to recognize that it is there.
+                    TypedReference reference = __makeref(t);
+                    ulong extremelyUnsafeAddress = **(ulong**)&reference;
```
Comment (arekpalinski) [inline bench/Micro.Benchmark/Benchmarks/TypeCache.cs:210]: Same here

File: src/Sparrow/Json/AbstractBlittableJsonTextWriter.cs [hunk 4]
```diff
@@ -214,8 +214,25 @@ public int WriteDateTime(DateTime value, bool isUtc)
...
+            if (str == null)
+            {
+                WriteNull();
+                return;
+            }
+
+            if (str.Length == 1)
+            {
+                char b = str[0];
+                if (b < 32 || (b >= 127 && b <= 159))  // is a control code character?
```
Comment (arekpalinski) [inline src/Sparrow/Json/AbstractBlittableJsonTextWriter.cs]: Can we have `32`, `127` and `159` defined as constants with meaningful names?

Comment (arekpalinski) [reply src/Sparrow/Json/AbstractBlittableJsonTextWriter.cs]: Or can we use `LazyStringValue.IsControlCodeCharacter()` here? (maybe make it static?)

**Comment by ppekrol** [comment]
There are at least 2 related failing tests, please check them

File: src/Sparrow/Utils/WeakTypeCache.cs (hunk 1 above)
Comment (redknightlois) [reply src/Sparrow/Utils/WeakTypeCache.cs]: No, it was moved to Micro.Benchmark. No use for it, removed.

File: bench/Micro.Benchmark/Benchmarks/TypeCache.cs (hunk 2 above)
Comment (redknightlois) [reply bench/Micro.Benchmark/Benchmarks/TypeCache.cs:171]: The warning is correct, that it is exactly what it is happening. And we are using the property that Type objects have a special treatment by the runtime. Where is that warning coming from as the CSC is not showing it inside Visual Studio?

File: src/Sparrow/Json/AbstractBlittableJsonTextWriter.cs (hunk 4 above)
Comment (redknightlois) [reply src/Sparrow/Json/AbstractBlittableJsonTextWriter.cs]: I ended up reverting it, there is no good way (effective) to achieve what I wanted to achieve without tripping on one of the errors showcase at the Slow Tests. I will need to revisit (eventually) LazyString now that we have Span.

**Review by arekpalinski — APPROVED** [review]

**Pull request merged**